### DISCUSSION TITLE
Checking for existence of local file or directory

### DISCUSSION THREAD

**Discussion by djarnis73** [apple/pkl#806] [Q&A]
It would be nice to be able to check for existence of a file or directory as a type constraint check.

I was thinking something in the line of:

```
class Container {
  dir: String(isDirectory)
  file: String(exist(dir, this))
}
```

My use case is that I try to build a bunch of deployment config files (mix of kubernetes/drone) where some values point to files/directories where it would be nice to tell the user if file/dir could not be found.

**Comment by StefMa** [comment]
You probably can make use of `read()` and `read?()`?!
https://pkl-lang.org/main/current/language-reference/index.html#resources

`exist()` is kind already implemented with `read()`, as it throws if it doesn't exist.
Alternatively, in case you want to throw another exception or using a fallback, you can use `read?() ?? throw("File does not exist")`

`isDirectory` could maybe somehow faked with `read*()` 🤔 

```pkl
class Container {
    dir: String(read*(this).keys.length > 0)
    file: String(read?(dir + "/" + this) != null)
}

container: Container = new {
    dir = "../workflows"
    file = "x.pkl"
}

```

**Reply by holzensp** [reply]
I agree with @StefMa; you can already implement `exists` in terms of `read?`;
```
fileExists = (it) -> read?(it) != null
```

There is no real `isDirectory` or equivalent. That said, when you `read` a directory, the `text` in the resulting `Resource` (which doesn't have `keys`, btw) is the listing of the directory's contents. So, you could...
```
class Container {
  function fileExists(it: String) = read?(it) != null
  dir: String(fileExists(this))
  local dirListing = read(dir).text.split("\n")
  file: String(dirListing.contains(this), fileExists(dir + "/" + this))
}
```
^^ There's redundancy there.

**Comment by djarnis73** [comment]
Thanks for the pointers. I experimented a bit. I could not get the "directory exist" feature to work (I did not try very hard so it is probably possible), but the following typealias works nice with files:

```
typealias LocalFile = String((it) -> (read?("file:" + it) ?? throw("File '\(this)' does not exist")) is Any)
```
I can use it like:
```
class Container {
  dir: String = "defaultDir"
  dockerfile: String = "Dockerfile"
  fixed dockerfileWithPath: LocalFile = dir + "/" + dockerfile
}

myContainer: Container
```
And it will fail with:

```
–– Pkl Error ––
File 'defaultDir/Dockerfile' does not exist

4 | typealias LocalFile = String((it) -> (read?("file:" + it) ?? throw("File '\(this)' does not exist")) is Any)
                                                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
at FileExists#LocalFile.<function#1> (file:.../FileExists.pkl, line 4)

10 | fixed fullDockerFile: LocalFile = dir + "/" + dockerfile
                                       ^^^^^^^^^^^^^^^^^^^^^^
at FileExists#Container.fullDockerFile (file:///.../FileExists.pkl, line 10)

106 | text = renderer.renderDocument(value)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
at pkl.base#Module.output.text (https://github.com/apple/pkl/blob/0.27.0/stdlib/base.pkl#L106)
```
If the file does not exist locally. This is good enough for my current needs.

One tiny annoyance is that the intellij plugin is clever enough to determine that the expression `read?("file:" + it) ?? throw("File '\(this)' does not exist")) is Any` is always true, so it highlights it as a warning, not sure if there is a way around that?
